File: alligator/src/main/java/me/aartikov/alligator/navigators/DefaultDialogFragmentNavigator.java
package me.aartikov.alligator.navigators;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.DialogFragment;
import androidx.fragment.app.FragmentManager;

import me.aartikov.alligator.Screen;
import me.aartikov.alligator.ScreenResult;
import me.aartikov.alligator.animations.AnimationData;
import me.aartikov.alligator.animations.DialogAnimation;
import me.aartikov.alligator.animations.providers.DialogAnimationProvider;
import me.aartikov.alligator.destinations.DialogFragmentDestination;
import me.aartikov.alligator.exceptions.NavigationException;
import me.aartikov.alligator.helpers.DialogFragmentHelper;
import me.aartikov.alligator.helpers.ScreenResultHelper;
import me.aartikov.alligator.listeners.DialogShowingListener;
import me.aartikov.alligator.listeners.ScreenResultListener;
import me.aartikov.alligator.navigationfactories.NavigationFactory;

public class DefaultDialogFragmentNavigator implements DialogFragmentNavigator {
	@NonNull
	private DialogFragmentHelper mDialogFragmentHelper;

	@NonNull
	private NavigationFactory mNavigationFactory;

	@NonNull
	private ScreenResultHelper mScreenResultHelper;

	@NonNull
	private DialogShowingListener mDialogShowingListener;

	@NonNull
	private ScreenResultListener mScreenResultListener;

	@NonNull
	private DialogAnimationProvider mAnimationProvider;

	public DefaultDialogFragmentNavigator(@NonNull FragmentManager fragmentManager,
										  @NonNull NavigationFactory navigationFactory,
										  @NonNull DialogShowingListener dialogShowingListener,
										  @NonNull ScreenResultListener screenResultListener,
										  @NonNull DialogAnimationProvider animationProvider) {
		mDialogFragmentHelper = new DialogFragmentHelper(fragmentManager);
		mNavigationFactory = navigationFactory;
		mScreenResultHelper = new ScreenResultHelper();
		mDialogShowingListener = dialogShowingListener;
		mScreenResultListener = screenResultListener;
		mAnimationProvider = animationProvider;
	}

	@Override
	public void goForward(@NonNull Screen screen,
						  @NonNull DialogFragmentDestination destination,
						  @Nullable AnimationData animationData) throws NavigationException {

		DialogFragment dialogFragment = destination.createDialogFragment(screen);
		DialogAnimation animation = mAnimationProvider.getAnimation(screen.getClass(), animationData);
		mDialogFragmentHelper.showDialog(dialogFragment, animation);
		mDialogShowingListener.onDialogShown(screen.getClass());
	}

	@Override
	public void replace(@NonNull Screen screen,
						@NonNull DialogFragmentDestination destination,
						@Nullable AnimationData animationData) throws NavigationException {
		if (mDialogFragmentHelper.isDialogVisible()) {
			mDialogFragmentHelper.hideDialog();
		}

		DialogFragment dialogFragment = destination.createDialogFragment(screen);
		DialogAnimation animation = mAnimationProvider.getAnimation(screen.getClass(), animationData);
		mDialogFragmentHelper.showDialog(dialogFragment, animation);
		mDialogShowingListener.onDialogShown(screen.getClass());
	}

	@Override
	public void reset(@NonNull Screen screen,
					  @NonNull DialogFragmentDestination destination,
					  @Nullable AnimationData animationData) throws NavigationException {

		while (mDialogFragmentHelper.isDialogVisible()) {
			mDialogFragmentHelper.hideDialog();
		}

		DialogFragment dialogFragment = destination.createDialogFragment(screen);
		DialogAnimation animation = mAnimationProvider.getAnimation(screen.getClass(), animationData);
		mDialogFragmentHelper.showDialog(dialogFragment, animation);
		mDialogShowingListener.onDialogShown(screen.getClass());
	}

	@Override
	public boolean canGoBack() {
		return mDialogFragmentHelper.isDialogVisible();
	}

	@Override
	public void goBack(@Nullable ScreenResult screenResult) throws NavigationException {
		DialogFragment dialogFragment = mDialogFragmentHelper.getDialogFragment();
		mDialogFragmentHelper.hideDialog();
		if (dialogFragment != null) {
			mScreenResultHelper.callScreenResultListener(dialogFragment, screenResult, mScreenResultListener, mNavigationFactory);
		}
	}
}

File: alligator/src/main/java/me/aartikov/alligator/navigators/DefaultDialogFragmentNavigator.kt
package me.aartikov.alligator.navigators

import androidx.fragment.app.DialogFragment
import androidx.fragment.app.FragmentManager
import me.aartikov.alligator.Screen
import me.aartikov.alligator.ScreenResult
import me.aartikov.alligator.animations.AnimationData
import me.aartikov.alligator.animations.providers.DialogAnimationProvider
import me.aartikov.alligator.destinations.DialogFragmentDestination
import me.aartikov.alligator.exceptions.NavigationException
import me.aartikov.alligator.helpers.DialogFragmentHelper
import me.aartikov.alligator.helpers.ScreenResultHelper
import me.aartikov.alligator.listeners.DialogShowingListener
import me.aartikov.alligator.listeners.ScreenResultListener
import me.aartikov.alligator.navigationfactories.NavigationFactory

class DefaultDialogFragmentNavigator(
    fragmentManager: FragmentManager,
    navigationFactory: NavigationFactory,
    dialogShowingListener: DialogShowingListener,
    screenResultListener: ScreenResultListener,
    animationProvider: DialogAnimationProvider
) : DialogFragmentNavigator {
    private val mDialogFragmentHelper: DialogFragmentHelper
    private val mNavigationFactory: NavigationFactory
    private val mScreenResultHelper: ScreenResultHelper
    private val mDialogShowingListener: DialogShowingListener
    private val mScreenResultListener: ScreenResultListener
    private val mAnimationProvider: DialogAnimationProvider

    init {
        mDialogFragmentHelper = DialogFragmentHelper(fragmentManager)
        mNavigationFactory = navigationFactory
        mScreenResultHelper = ScreenResultHelper(mNavigationFactory)
        mDialogShowingListener = dialogShowingListener
        mScreenResultListener = screenResultListener
        mAnimationProvider = animationProvider
    }

    @Throws(NavigationException::class)
    override fun goForward(
        screen: Screen,
        destination: DialogFragmentDestination,
        animationData: AnimationData?
    ) {
        val dialogFragment = destination.createDialogFragment(screen)
        val animation = mAnimationProvider.getAnimation(screen.javaClass, animationData)
        mDialogFragmentHelper.showDialog(dialogFragment, animation)
        mDialogShowingListener.onDialogShown(screen.javaClass)
    }

    @Throws(NavigationException::class)
    override fun replace(
        screen: Screen,
        destination: DialogFragmentDestination,
        animationData: AnimationData?
    ) {
        if (mDialogFragmentHelper.isDialogVisible) {
            mDialogFragmentHelper.hideDialog()
        }
        val dialogFragment = destination.createDialogFragment(screen)
        val animation = mAnimationProvider.getAnimation(screen.javaClass, animationData)
        mDialogFragmentHelper.showDialog(dialogFragment, animation)
        mDialogShowingListener.onDialogShown(screen.javaClass)
    }

    @Throws(NavigationException::class)
    override fun reset(
        screen: Screen,
        destination: DialogFragmentDestination,
        animationData: AnimationData?
    ) {
        while (mDialogFragmentHelper.isDialogVisible) {
            mDialogFragmentHelper.hideDialog()
        }
        val dialogFragment = destination.createDialogFragment(screen)
        val animation = mAnimationProvider.getAnimation(screen.javaClass, animationData)
        mDialogFragmentHelper.showDialog(dialogFragment, animation)
        mDialogShowingListener.onDialogShown(screen.javaClass)
    }

    override fun canGoBack(): Boolean {
        return mDialogFragmentHelper.isDialogVisible
    }

    @Throws(NavigationException::class)
    override fun goBack(screenResult: ScreenResult?) {
        val dialogFragment = mDialogFragmentHelper.dialogFragment
        mDialogFragmentHelper.hideDialog()
        if (dialogFragment != null) {
            mScreenResultHelper.callScreenResultListener(
                dialogFragment,
                screenResult,
                mScreenResultListener
            )
        }
    }

    override val currentDialogFragment: DialogFragment?
        get() = mDialogFragmentHelper.dialogFragment
}